ddtrace/appsec/_common_module_patches.py

```diff
@@ -0,0 +1,134 @@
+import ctypes
+import gc
+from typing import Any
+from typing import Callable
+from typing import Dict
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.settings.asm import config as asm_config
+from ddtrace.vendor.wrapt import FunctionWrapper
+from ddtrace.vendor.wrapt import resolve_path
+
+
+log = get_logger(__name__)
+_DD_ORIGINAL_ATTRIBUTES: Dict[Any, Any] = {}
+
+
+def patch_common_modules():
+    try_wrap_function_wrapper("builtins", "open", wrapped_open_CFDDB7ABBA9081B6)
+
+
+def unpatch_common_modules():
+    try_unwrap("builtins", "open")
+
+
+def wrapped_open_CFDDB7ABBA9081B6(original_open_callable, instance, args, kwargs):
+    if asm_config._iast_enabled:
+        # LFI sink to be added
+        pass
+
+    if asm_config._asm_enabled and asm_config._ep_enabled:
+        try:
+            from ddtrace.appsec._asm_request_context import call_waf_callback
+            from ddtrace.appsec._asm_request_context import in_context
+        except ImportError:
+            # open is used during module initialization
+            # and shouldn't be changed at that time
+            return original_open_callable(*args, **kwargs)
+
+        if len(args) > 0 and in_context():
+            call_waf_callback({"LFI_ADDRESS": args[0]}, crop_trace="wrapped_open_CFDDB7ABBA9081B6")
+            # DEV: Next part of the exploit prevention feature: add block here
+    return original_open_callable(*args, **kwargs)
+
+
+def try_unwrap(module, name):
+    try:
+        (parent, attribute, _) = resolve_path(module, name)
+        if (parent, attribute) in _DD_ORIGINAL_ATTRIBUTES:
+            original = _DD_ORIGINAL_ATTRIBUTES[(parent, attribute)]
+            apply_patch(parent, attribute, original)
+            del _DD_ORIGINAL_ATTRIBUTES[(parent, attribute)]
+    except ModuleNotFoundError:
+        pass
+
+
+def try_wrap_function_wrapper(module: str, name: str, wrapper: Callable) -> None:
+    try:
+        wrap_object(module, name, FunctionWrapper, (wrapper,))
+    except (ImportError, AttributeError):
+        log.debug("ASM patching. Module %s.%s does not exist", module, name)
+
+
+def wrap_object(module, name, factory, args=(), kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    (parent, attribute, original) = resolve_path(module, name)
+    wrapper = factory(original, *args, **kwargs)
+    apply_patch(parent, attribute, wrapper)
+    return wrapper
+
+
+def apply_patch(parent, attribute, replacement):
+    try:
+        current_attribute = getattr(parent, attribute)
+        # Avoid overwriting the original function if we call this twice
```

Comment:
We do this in other parts of the tracer setting/getting a `_datadog_patch` attribute
https://github.com/DataDog/dd-trace-py/blob/3d1e9c9960b351fbf1af0e218d0e88ffdde9f5ec/ddtrace/contrib/flask/patch.py#L165-L167

Reply:
Not sure it applies here though

Reply:
Again just moved that code from ddtrace/appsec/_iast/_patch.py